tests/unit_tests/dft/include/test_common.hpp

```diff
@@ -0,0 +1,197 @@
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+#ifndef ONEMKL_TEST_COMMON_HPP
+#define ONEMKL_TEST_COMMON_HPP
+
+#include <cstdint>
+#include <cstdlib>
+#include <iostream>
+#include <limits>
+#include <vector>
+
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
+
+template <typename T>
+struct complex_info {
+    using real_type = T;
+    static const bool is_complex = false;
+};
+
+template <typename T>
+struct complex_info<std::complex<T>> {
+    using real_type = T;
+    static const bool is_complex = true;
+};
+
+template <typename T>
+constexpr bool is_complex() {
+    return complex_info<T>::is_complex;
+}
+template <typename T>
+constexpr int num_components() {
+    return is_complex<T>() ? 2 : 1;
+}
+
+template <typename fp>
+bool check_equal(fp x, fp x_ref, int error_mag) {
+    using fp_real = typename complex_info<fp>::real_type;
+    static_assert(std::is_floating_point_v<fp_real>,
+                  "Expected floating-point real or complex type.");
+
+    const fp_real bound = [error_mag]() {
+        if constexpr (sizeof(double) == sizeof(long double) && std::is_same_v<fp_real, double>) {
+            // The reference DFT uses long double to maintain accuracy
+            // when this isn't possible, lower the accuracy requirements
+            return 1e-12;
```

Comment:
Is this absolute/relative error threshold for double-precision to be satisfied for real and complex `fp` type? Also, why such a distinction between single and double precision?

Reply:
One of the challenges in the reference DFT was accuracy. Originally, we computed the reference DFT results in double precision. This meant that we could get an accurate result for single precision, and but there was numerical error introduced in the reference double-precision result. This error in the double-precision reference required us to set a larger error margin than we would have liked. The solution to this is to use `long double` for computing the reference. 

However, there is a problem in our solution: `long double` is equivalent to `double` on some compilers. Consequently, we may not always be able to use the extended precision to compute more accurate reference results for double precision. The solution to this is to allow increased error margins when using `long double` is `double` and we are checking the accuracy of double-precision results.

Yes, this is an error threshold for both real and complex types.

Reply:
Speaking of accuracy of the reference impl, I think we would probably still have issues with a naive implementation like this if the FFT size is large enough. Whether that matters depends on the purpose of providing these unit tests. I've assumed the purpose is to add a few checks for small sizes mostly to guarantee the interface layer between these oneMKL APIs and the backend is functioning, and that any backend should have its own much larger suite of tests that would cover accuracy at larger sizes. Is that assumption correct?